{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "polish-henry",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import h5py\n",
    "from matplotlib import pyplot as plt\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.linear_model import Lasso, LogisticRegression\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "from scipy.stats import spearmanr\n",
    "\n",
    "from tensorflow import keras\n",
    "from keras import layers\n",
    "\n",
    "import MMRF_utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "specified-carol",
   "metadata": {},
   "outputs": [],
   "source": [
    "mmrf_rna_dir = '/Users/nalinisingh/datasets/multiple_myeloma/MMRF_CoMMpass_IA15a_E74GTF_Cufflinks_Gene_FPKM.txt'\n",
    "asct_data_dir = '/Users/nalinisingh/datasets/multiple_myeloma/MMRF_OS_PFS_ASCT.csv'\n",
    "no_asct_data_dir = '/Users/nalinisingh/datasets/multiple_myeloma/MMRF_OS_PFS_non-ASCT.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "periodic-perth",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loading from: /Users/nalinisingh/dev/ml_mmrf_nonnormalized/ml_mmrf/ml_mmrf/output/cleaned_mm_fold_2mos_pfs_ind_seed0.pkl\n",
      "pids\n",
      "(470,)\n",
      "x\n",
      "(470, 33, 16)\n",
      "m\n",
      "(470, 33, 16)\n",
      "feature_names_x\n",
      "(16,)\n",
      "['cbc_abs_neut' 'chem_albumin' 'chem_bun' 'chem_calcium' 'chem_creatinine'\n",
      " 'chem_glucose' 'cbc_hemoglobin' 'serum_kappa' 'serum_m_protein'\n",
      " 'cbc_platelet' 'chem_totprot' 'cbc_wbc' 'serum_iga' 'serum_igg'\n",
      " 'serum_igm' 'serum_lambda']\n",
      "ys_seq\n",
      "(470, 1)\n",
      "ce\n",
      "(470, 1)\n",
      "feature_names_y\n",
      "(1,)\n",
      "['progression free survival (all)']\n",
      "b\n",
      "(470, 17)\n",
      "feature_names\n",
      "(17,)\n",
      "Index(['iss', 'age', 'gender', 'ecog', 'line1sct', 'serum_beta2_microglobulin',\n",
      "       'PC1', 'PC2', 'PC3', 'PC4', 'PC5', 'heavy_chain', 'igg_type',\n",
      "       'iga_type', 'igm_type', 'kappa_type', 'lambda_type'],\n",
      "      dtype='object')\n",
      "a\n",
      "(470, 33, 9)\n",
      "m_a\n",
      "(470, 33, 6)\n",
      "feature_names_a\n",
      "(9,)\n",
      "['local_clock' 'Bor' 'Car' 'Cyc' 'Dex' 'Len' 'line1' 'line2' 'line3plus']\n",
      "Keep first 3 clinical visits.\n",
      "Keep first 3 clinical visits.\n",
      "Keep first 3 clinical visits.\n",
      "Reading /Users/nalinisingh/datasets/multiple_myeloma/MMRF_CoMMpass_IA15a_E74GTF_Salmon_Gene_TPM.txt\n",
      "Raw data shape:\n",
      "(57997, 922)\n",
      "0 out of 57997 genes have missing data \n",
      "Using raw genomic data...\n",
      "Created new dataframe...\n"
     ]
    }
   ],
   "source": [
    "train_patient_all = pd.DataFrame()\n",
    "valid_patient_all = pd.DataFrame()\n",
    "test_patient_all = pd.DataFrame()\n",
    "\n",
    "for ind in [1]:\n",
    "    data_filename = '/Users/nalinisingh/dev/ml_mmrf_nonnormalized/ml_mmrf/ml_mmrf/output/cleaned_mm_fold_2mos_pfs_ind_seed0.pkl'#%(ind)\n",
    "    train, test, valid = MMRF_utils.get_train_test_valid(data_filename, ind, show_features=True)\n",
    "\n",
    "    # include the first n clinical visit data for each patient (default 0)\n",
    "    train_df = MMRF_utils.preprocess_patient_data(train, num_clin_visits=3)  \n",
    "\n",
    "    valid_df = MMRF_utils.preprocess_patient_data(valid, num_clin_visits=3)  \n",
    "\n",
    "    test_df = MMRF_utils.preprocess_patient_data(test, num_clin_visits=3)  \n",
    "\n",
    "    genomic_fn = '/Users/nalinisingh/datasets/multiple_myeloma/MMRF_CoMMpass_IA15a_E74GTF_Salmon_Gene_TPM.txt'\n",
    "    genomic_df = MMRF_utils.preprocess_genomic_data(genomic_fn, nPCA=0)\n",
    "\n",
    "    train_patient_all = train_patient_all.append(train_df.merge(genomic_df, left_on='pids', right_on='pids'))\n",
    "\n",
    "    valid_patient_all = valid_patient_all.append(valid_df.merge(genomic_df, left_on='pids', right_on='pids'))\n",
    "\n",
    "    test_patient_all = test_patient_all.append(test_df.merge(genomic_df, left_on='pids', right_on='pids'))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 272,
   "id": "promotional-delta",
   "metadata": {},
   "outputs": [],
   "source": [
    "asct_data = pd.read_csv('/Users/nalinisingh/datasets/multiple_myeloma/MMRF_OS_PFS_ASCT.csv')\n",
    "no_asct_data = pd.read_csv('/Users/nalinisingh/datasets/multiple_myeloma/MMRF_OS_PFS_non-ASCT.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 273,
   "id": "infectious-metropolitan",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_all_patient = pd.concat([train_df, valid_df, test_df])\n",
    "combined_all_patient['gender'] = (combined_all_patient['gender']-1)*100\n",
    "combined_all_rna_seq = pd.concat([train_patient_all,valid_patient_all,test_patient_all])\n",
    "combined_all_rna_seq['gender'] = (combined_all_rna_seq['gender']-1)*100\n",
    "bort_rna_seq = combined_all_rna_seq[combined_all_rna_seq['Bor1']==True]\n",
    "asct_rna_seq = combined_all_rna_seq[combined_all_rna_seq['pids'].isin(asct_data['Patient'].values)]\n",
    "no_asct_rna_seq = combined_all_rna_seq[combined_all_rna_seq['pids'].isin(no_asct_data['Patient'].values)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 320,
   "id": "known-evening",
   "metadata": {},
   "outputs": [],
   "source": [
    "covs = ['age','gender','iss','ecog','igg_type','igm_type','iga_type','kappa_type','lambda_type','serum_beta2_microglobulin','serum_m_protein1','chem_creatinine1','chem_albumin1','pfs']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 353,
   "id": "beautiful-brazilian",
   "metadata": {},
   "outputs": [],
   "source": [
    "cohort_mean_df = pd.DataFrame()\n",
    "cohort_std_df = pd.DataFrame()\n",
    "\n",
    "for i, df in enumerate([combined_all_patient, combined_all_rna_seq, bort_rna_seq, asct_rna_seq, no_asct_rna_seq]):\n",
    "    cohort_mean_df = cohort_mean_df.append(pd.DataFrame(df[covs].mean()).T.applymap(lambda x: '{0:.2f}'.format(x)),ignore_index=True)\n",
    "    cohort_std_df = cohort_std_df.append(pd.DataFrame(df[covs].std()).T.applymap(lambda x: '{0:.2f}'.format(x)),ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "id": "saving-courage",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>age</th>\n",
       "      <th>gender</th>\n",
       "      <th>iss</th>\n",
       "      <th>ecog</th>\n",
       "      <th>igg_type</th>\n",
       "      <th>igm_type</th>\n",
       "      <th>iga_type</th>\n",
       "      <th>kappa_type</th>\n",
       "      <th>lambda_type</th>\n",
       "      <th>serum_beta2_microglobulin</th>\n",
       "      <th>serum_m_protein1</th>\n",
       "      <th>chem_creatinine1</th>\n",
       "      <th>chem_albumin1</th>\n",
       "      <th>pfs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>63.31</td>\n",
       "      <td>40.10</td>\n",
       "      <td>1.85</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.65</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.17</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.34</td>\n",
       "      <td>4.85</td>\n",
       "      <td>2.81</td>\n",
       "      <td>113.94</td>\n",
       "      <td>35.96</td>\n",
       "      <td>16.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>63.22</td>\n",
       "      <td>41.09</td>\n",
       "      <td>1.85</td>\n",
       "      <td>0.88</td>\n",
       "      <td>0.65</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.35</td>\n",
       "      <td>4.91</td>\n",
       "      <td>2.85</td>\n",
       "      <td>112.90</td>\n",
       "      <td>35.93</td>\n",
       "      <td>16.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>63.55</td>\n",
       "      <td>39.65</td>\n",
       "      <td>1.87</td>\n",
       "      <td>0.88</td>\n",
       "      <td>0.62</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.34</td>\n",
       "      <td>5.13</td>\n",
       "      <td>2.80</td>\n",
       "      <td>118.21</td>\n",
       "      <td>35.93</td>\n",
       "      <td>16.91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>59.49</td>\n",
       "      <td>43.68</td>\n",
       "      <td>1.79</td>\n",
       "      <td>0.76</td>\n",
       "      <td>0.65</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.36</td>\n",
       "      <td>4.45</td>\n",
       "      <td>2.73</td>\n",
       "      <td>105.25</td>\n",
       "      <td>36.93</td>\n",
       "      <td>18.95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>67.77</td>\n",
       "      <td>37.92</td>\n",
       "      <td>1.94</td>\n",
       "      <td>1.03</td>\n",
       "      <td>0.66</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.65</td>\n",
       "      <td>0.33</td>\n",
       "      <td>5.48</td>\n",
       "      <td>2.99</td>\n",
       "      <td>122.24</td>\n",
       "      <td>34.70</td>\n",
       "      <td>12.48</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     age gender   iss  ecog igg_type igm_type iga_type kappa_type lambda_type  \\\n",
       "0  63.31  40.10  1.85  0.86     0.65     0.06     0.17       0.64        0.34   \n",
       "1  63.22  41.09  1.85  0.88     0.65     0.05     0.18       0.64        0.35   \n",
       "2  63.55  39.65  1.87  0.88     0.62     0.05     0.19       0.64        0.34   \n",
       "3  59.49  43.68  1.79  0.76     0.65     0.04     0.18       0.64        0.36   \n",
       "4  67.77  37.92  1.94  1.03     0.66     0.06     0.18       0.65        0.33   \n",
       "\n",
       "  serum_beta2_microglobulin serum_m_protein1 chem_creatinine1 chem_albumin1  \\\n",
       "0                      4.85             2.81           113.94         35.96   \n",
       "1                      4.91             2.85           112.90         35.93   \n",
       "2                      5.13             2.80           118.21         35.93   \n",
       "3                      4.45             2.73           105.25         36.93   \n",
       "4                      5.48             2.99           122.24         34.70   \n",
       "\n",
       "     pfs  \n",
       "0  16.29  \n",
       "1  16.04  \n",
       "2  16.91  \n",
       "3  18.95  \n",
       "4  12.48  "
      ]
     },
     "execution_count": 354,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cohort_mean_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "id": "recognized-tuesday",
   "metadata": {},
   "outputs": [],
   "source": [
    "cohort_str_df = pd.DataFrame(index= cohort_mean_df.index,columns=cohort_mean_df.columns)\n",
    "for i in range(5):\n",
    "\n",
    "    cohort_str_df.iloc[i] = cohort_mean_df.iloc[i]+r'$\\pm$'+cohort_std_df.iloc[i]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 356,
   "id": "several-conducting",
   "metadata": {},
   "outputs": [],
   "source": [
    "cohort_list = [\n",
    "    r'\\makecell{All \\\\ (n='+str(combined_all_patient.shape[0])+')}',\n",
    "    r'\\makecell{RNA-Seq \\\\ (n='+str(combined_all_rna_seq.shape[0])+')}',\n",
    "    r'\\makecell{Bortezomib \\\\ (n='+str(bort_rna_seq.shape[0])+')}',\n",
    "    r'\\makecell{ASCT \\\\ (n='+str(asct_rna_seq.shape[0])+')}',\n",
    "    r'\\makecell{Non-ASCT \\\\ (n='+str(no_asct_rna_seq.shape[0])+')}'\n",
    "]\n",
    "cohort_str_df = cohort_str_df.rename({i:cohort_list[i] for i in range(5)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 357,
   "id": "tight-bryan",
   "metadata": {},
   "outputs": [],
   "source": [
    "att_map = {\n",
    "    r'age': r'\\makecell{Age \\\\ (years)}',\n",
    "    r'gender': r'\\makecell{Gender \\\\ (\\% female)}',\n",
    "    r'iss': r'ISS',\n",
    "    r'ecog': r'ECOG',\n",
    "    r'igg_type': r'\\makecell{IgG \\\\ (g/L)}',\n",
    "    r'iga_type': r'\\makecell{IgA \\\\ (g/L)}',\n",
    "    r'igm_type': r'\\makecell{IgM \\\\ (g/L)}',\n",
    "    r'kappa_type': r'\\makecell{$\\kappa$-Light Chain \\\\ (mg/dL)}',\n",
    "    r'lambda_type': r'\\makecell{$\\lambda$-Light Chain \\\\ (mg/dL)}',\n",
    "    r'serum_beta2_microglobulin': r'\\makecell{$\\beta_2$-Microglobulin \\\\ (mg/mL)}',\n",
    "    r'serum_m_protein1': r'\\makecell{M-protein \\\\ (g/dL)}',\n",
    "    r'chem_creatinine1': r'\\makecell{Creatinine \\\\ ($\\mu$mol/L)}',\n",
    "    r'chem_albumin1': r'\\makecell{Albumin \\\\ (g/L)}',\n",
    "    r'pfs': r'PFS (months)'\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 358,
   "id": "prospective-grenada",
   "metadata": {},
   "outputs": [],
   "source": [
    "cohort_str_df = cohort_str_df.rename(att_map,axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 359,
   "id": "civil-possession",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>\\makecell{Age \\\\ (years)}</th>\n",
       "      <th>\\makecell{Gender \\\\ (\\% female)}</th>\n",
       "      <th>ISS</th>\n",
       "      <th>ECOG</th>\n",
       "      <th>\\makecell{IgG \\\\ (g/L)}</th>\n",
       "      <th>\\makecell{IgM \\\\ (g/L)}</th>\n",
       "      <th>\\makecell{IgA \\\\ (g/L)}</th>\n",
       "      <th>\\makecell{$\\kappa$-Light Chain \\\\ (mg/dL)}</th>\n",
       "      <th>\\makecell{$\\lambda$-Light Chain \\\\ (mg/dL)}</th>\n",
       "      <th>\\makecell{$\\beta_2$-Microglobulin \\\\ (mg/mL)}</th>\n",
       "      <th>\\makecell{M-protein \\\\ (g/dL)}</th>\n",
       "      <th>\\makecell{Creatinine \\\\ ($\\mu$mol/L)}</th>\n",
       "      <th>\\makecell{Albumin \\\\ (g/L)}</th>\n",
       "      <th>PFS (months)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>\\makecell{All \\\\ (n=995)}</th>\n",
       "      <td>63.31$\\pm$10.47</td>\n",
       "      <td>40.10$\\pm$49.03</td>\n",
       "      <td>1.85$\\pm$0.45</td>\n",
       "      <td>0.86$\\pm$0.73</td>\n",
       "      <td>0.65$\\pm$0.48</td>\n",
       "      <td>0.06$\\pm$0.23</td>\n",
       "      <td>0.17$\\pm$0.37</td>\n",
       "      <td>0.64$\\pm$0.48</td>\n",
       "      <td>0.34$\\pm$0.47</td>\n",
       "      <td>4.85$\\pm$3.92</td>\n",
       "      <td>2.81$\\pm$1.81</td>\n",
       "      <td>113.94$\\pm$74.25</td>\n",
       "      <td>35.96$\\pm$6.23</td>\n",
       "      <td>16.29$\\pm$9.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>\\makecell{RNA-Seq \\\\ (n=662)}</th>\n",
       "      <td>63.22$\\pm$10.76</td>\n",
       "      <td>41.09$\\pm$49.24</td>\n",
       "      <td>1.85$\\pm$0.49</td>\n",
       "      <td>0.88$\\pm$0.76</td>\n",
       "      <td>0.65$\\pm$0.48</td>\n",
       "      <td>0.05$\\pm$0.21</td>\n",
       "      <td>0.18$\\pm$0.38</td>\n",
       "      <td>0.64$\\pm$0.48</td>\n",
       "      <td>0.35$\\pm$0.48</td>\n",
       "      <td>4.91$\\pm$3.88</td>\n",
       "      <td>2.85$\\pm$1.77</td>\n",
       "      <td>112.90$\\pm$72.86</td>\n",
       "      <td>35.93$\\pm$6.27</td>\n",
       "      <td>16.04$\\pm$9.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>\\makecell{Bortezomib \\\\ (n=512)}</th>\n",
       "      <td>63.55$\\pm$10.62</td>\n",
       "      <td>39.65$\\pm$48.96</td>\n",
       "      <td>1.87$\\pm$0.49</td>\n",
       "      <td>0.88$\\pm$0.79</td>\n",
       "      <td>0.62$\\pm$0.49</td>\n",
       "      <td>0.05$\\pm$0.22</td>\n",
       "      <td>0.19$\\pm$0.40</td>\n",
       "      <td>0.64$\\pm$0.48</td>\n",
       "      <td>0.34$\\pm$0.48</td>\n",
       "      <td>5.13$\\pm$3.98</td>\n",
       "      <td>2.80$\\pm$1.80</td>\n",
       "      <td>118.21$\\pm$79.43</td>\n",
       "      <td>35.93$\\pm$6.37</td>\n",
       "      <td>16.91$\\pm$9.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>\\makecell{ASCT \\\\ (n=364)}</th>\n",
       "      <td>59.49$\\pm$9.18</td>\n",
       "      <td>43.68$\\pm$49.67</td>\n",
       "      <td>1.79$\\pm$0.52</td>\n",
       "      <td>0.76$\\pm$0.67</td>\n",
       "      <td>0.65$\\pm$0.48</td>\n",
       "      <td>0.04$\\pm$0.19</td>\n",
       "      <td>0.18$\\pm$0.38</td>\n",
       "      <td>0.64$\\pm$0.48</td>\n",
       "      <td>0.36$\\pm$0.48</td>\n",
       "      <td>4.45$\\pm$3.68</td>\n",
       "      <td>2.73$\\pm$1.79</td>\n",
       "      <td>105.25$\\pm$68.38</td>\n",
       "      <td>36.93$\\pm$6.18</td>\n",
       "      <td>18.95$\\pm$8.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>\\makecell{Non-ASCT \\\\ (n=298)}</th>\n",
       "      <td>67.77$\\pm$10.81</td>\n",
       "      <td>37.92$\\pm$48.60</td>\n",
       "      <td>1.94$\\pm$0.44</td>\n",
       "      <td>1.03$\\pm$0.83</td>\n",
       "      <td>0.66$\\pm$0.48</td>\n",
       "      <td>0.06$\\pm$0.24</td>\n",
       "      <td>0.18$\\pm$0.39</td>\n",
       "      <td>0.65$\\pm$0.48</td>\n",
       "      <td>0.33$\\pm$0.47</td>\n",
       "      <td>5.48$\\pm$4.04</td>\n",
       "      <td>2.99$\\pm$1.75</td>\n",
       "      <td>122.24$\\pm$77.09</td>\n",
       "      <td>34.70$\\pm$6.18</td>\n",
       "      <td>12.48$\\pm$8.56</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 \\makecell{Age \\\\ (years)}  \\\n",
       "\\makecell{All \\\\ (n=995)}                  63.31$\\pm$10.47   \n",
       "\\makecell{RNA-Seq \\\\ (n=662)}              63.22$\\pm$10.76   \n",
       "\\makecell{Bortezomib \\\\ (n=512)}           63.55$\\pm$10.62   \n",
       "\\makecell{ASCT \\\\ (n=364)}                  59.49$\\pm$9.18   \n",
       "\\makecell{Non-ASCT \\\\ (n=298)}             67.77$\\pm$10.81   \n",
       "\n",
       "                                 \\makecell{Gender \\\\ (\\% female)}  \\\n",
       "\\makecell{All \\\\ (n=995)}                         40.10$\\pm$49.03   \n",
       "\\makecell{RNA-Seq \\\\ (n=662)}                     41.09$\\pm$49.24   \n",
       "\\makecell{Bortezomib \\\\ (n=512)}                  39.65$\\pm$48.96   \n",
       "\\makecell{ASCT \\\\ (n=364)}                        43.68$\\pm$49.67   \n",
       "\\makecell{Non-ASCT \\\\ (n=298)}                    37.92$\\pm$48.60   \n",
       "\n",
       "                                            ISS           ECOG  \\\n",
       "\\makecell{All \\\\ (n=995)}         1.85$\\pm$0.45  0.86$\\pm$0.73   \n",
       "\\makecell{RNA-Seq \\\\ (n=662)}     1.85$\\pm$0.49  0.88$\\pm$0.76   \n",
       "\\makecell{Bortezomib \\\\ (n=512)}  1.87$\\pm$0.49  0.88$\\pm$0.79   \n",
       "\\makecell{ASCT \\\\ (n=364)}        1.79$\\pm$0.52  0.76$\\pm$0.67   \n",
       "\\makecell{Non-ASCT \\\\ (n=298)}    1.94$\\pm$0.44  1.03$\\pm$0.83   \n",
       "\n",
       "                                 \\makecell{IgG \\\\ (g/L)}  \\\n",
       "\\makecell{All \\\\ (n=995)}                  0.65$\\pm$0.48   \n",
       "\\makecell{RNA-Seq \\\\ (n=662)}              0.65$\\pm$0.48   \n",
       "\\makecell{Bortezomib \\\\ (n=512)}           0.62$\\pm$0.49   \n",
       "\\makecell{ASCT \\\\ (n=364)}                 0.65$\\pm$0.48   \n",
       "\\makecell{Non-ASCT \\\\ (n=298)}             0.66$\\pm$0.48   \n",
       "\n",
       "                                 \\makecell{IgM \\\\ (g/L)}  \\\n",
       "\\makecell{All \\\\ (n=995)}                  0.06$\\pm$0.23   \n",
       "\\makecell{RNA-Seq \\\\ (n=662)}              0.05$\\pm$0.21   \n",
       "\\makecell{Bortezomib \\\\ (n=512)}           0.05$\\pm$0.22   \n",
       "\\makecell{ASCT \\\\ (n=364)}                 0.04$\\pm$0.19   \n",
       "\\makecell{Non-ASCT \\\\ (n=298)}             0.06$\\pm$0.24   \n",
       "\n",
       "                                 \\makecell{IgA \\\\ (g/L)}  \\\n",
       "\\makecell{All \\\\ (n=995)}                  0.17$\\pm$0.37   \n",
       "\\makecell{RNA-Seq \\\\ (n=662)}              0.18$\\pm$0.38   \n",
       "\\makecell{Bortezomib \\\\ (n=512)}           0.19$\\pm$0.40   \n",
       "\\makecell{ASCT \\\\ (n=364)}                 0.18$\\pm$0.38   \n",
       "\\makecell{Non-ASCT \\\\ (n=298)}             0.18$\\pm$0.39   \n",
       "\n",
       "                                 \\makecell{$\\kappa$-Light Chain \\\\ (mg/dL)}  \\\n",
       "\\makecell{All \\\\ (n=995)}                                     0.64$\\pm$0.48   \n",
       "\\makecell{RNA-Seq \\\\ (n=662)}                                 0.64$\\pm$0.48   \n",
       "\\makecell{Bortezomib \\\\ (n=512)}                              0.64$\\pm$0.48   \n",
       "\\makecell{ASCT \\\\ (n=364)}                                    0.64$\\pm$0.48   \n",
       "\\makecell{Non-ASCT \\\\ (n=298)}                                0.65$\\pm$0.48   \n",
       "\n",
       "                                 \\makecell{$\\lambda$-Light Chain \\\\ (mg/dL)}  \\\n",
       "\\makecell{All \\\\ (n=995)}                                      0.34$\\pm$0.47   \n",
       "\\makecell{RNA-Seq \\\\ (n=662)}                                  0.35$\\pm$0.48   \n",
       "\\makecell{Bortezomib \\\\ (n=512)}                               0.34$\\pm$0.48   \n",
       "\\makecell{ASCT \\\\ (n=364)}                                     0.36$\\pm$0.48   \n",
       "\\makecell{Non-ASCT \\\\ (n=298)}                                 0.33$\\pm$0.47   \n",
       "\n",
       "                                 \\makecell{$\\beta_2$-Microglobulin \\\\ (mg/mL)}  \\\n",
       "\\makecell{All \\\\ (n=995)}                                        4.85$\\pm$3.92   \n",
       "\\makecell{RNA-Seq \\\\ (n=662)}                                    4.91$\\pm$3.88   \n",
       "\\makecell{Bortezomib \\\\ (n=512)}                                 5.13$\\pm$3.98   \n",
       "\\makecell{ASCT \\\\ (n=364)}                                       4.45$\\pm$3.68   \n",
       "\\makecell{Non-ASCT \\\\ (n=298)}                                   5.48$\\pm$4.04   \n",
       "\n",
       "                                 \\makecell{M-protein \\\\ (g/dL)}  \\\n",
       "\\makecell{All \\\\ (n=995)}                         2.81$\\pm$1.81   \n",
       "\\makecell{RNA-Seq \\\\ (n=662)}                     2.85$\\pm$1.77   \n",
       "\\makecell{Bortezomib \\\\ (n=512)}                  2.80$\\pm$1.80   \n",
       "\\makecell{ASCT \\\\ (n=364)}                        2.73$\\pm$1.79   \n",
       "\\makecell{Non-ASCT \\\\ (n=298)}                    2.99$\\pm$1.75   \n",
       "\n",
       "                                 \\makecell{Creatinine \\\\ ($\\mu$mol/L)}  \\\n",
       "\\makecell{All \\\\ (n=995)}                             113.94$\\pm$74.25   \n",
       "\\makecell{RNA-Seq \\\\ (n=662)}                         112.90$\\pm$72.86   \n",
       "\\makecell{Bortezomib \\\\ (n=512)}                      118.21$\\pm$79.43   \n",
       "\\makecell{ASCT \\\\ (n=364)}                            105.25$\\pm$68.38   \n",
       "\\makecell{Non-ASCT \\\\ (n=298)}                        122.24$\\pm$77.09   \n",
       "\n",
       "                                 \\makecell{Albumin \\\\ (g/L)}    PFS (months)  \n",
       "\\makecell{All \\\\ (n=995)}                     35.96$\\pm$6.23  16.29$\\pm$9.60  \n",
       "\\makecell{RNA-Seq \\\\ (n=662)}                 35.93$\\pm$6.27  16.04$\\pm$9.31  \n",
       "\\makecell{Bortezomib \\\\ (n=512)}              35.93$\\pm$6.37  16.91$\\pm$9.02  \n",
       "\\makecell{ASCT \\\\ (n=364)}                    36.93$\\pm$6.18  18.95$\\pm$8.89  \n",
       "\\makecell{Non-ASCT \\\\ (n=298)}                34.70$\\pm$6.18  12.48$\\pm$8.56  "
      ]
     },
     "execution_count": 359,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cohort_str_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 361,
   "id": "champion-stanley",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\\begin{tabular}{cccccc}\n",
      "\\toprule\n",
      "{} & \\makecell{All \\\\ (n=995)} & \\makecell{RNA-Seq \\\\ (n=662)} & \\makecell{Bortezomib \\\\ (n=512)} & \\makecell{ASCT \\\\ (n=364)} & \\makecell{Non-ASCT \\\\ (n=298)} \\\\ \\hline\n",
      "\\midrule\n",
      "\\makecell{Age \\\\ (years)}                     &           63.31$\\pm$10.47 &               63.22$\\pm$10.76 &                  63.55$\\pm$10.62 &             59.49$\\pm$9.18 &                67.77$\\pm$10.81 \\\\ \\hline\n",
      "\\makecell{Gender \\\\ (\\% female)}              &           40.10$\\pm$49.03 &               41.09$\\pm$49.24 &                  39.65$\\pm$48.96 &            43.68$\\pm$49.67 &                37.92$\\pm$48.60 \\\\ \\hline\n",
      "ISS                                           &             1.85$\\pm$0.45 &                 1.85$\\pm$0.49 &                    1.87$\\pm$0.49 &              1.79$\\pm$0.52 &                  1.94$\\pm$0.44 \\\\ \\hline\n",
      "ECOG                                          &             0.86$\\pm$0.73 &                 0.88$\\pm$0.76 &                    0.88$\\pm$0.79 &              0.76$\\pm$0.67 &                  1.03$\\pm$0.83 \\\\ \\hline\n",
      "\\makecell{IgG \\\\ (g/L)}                       &             0.65$\\pm$0.48 &                 0.65$\\pm$0.48 &                    0.62$\\pm$0.49 &              0.65$\\pm$0.48 &                  0.66$\\pm$0.48 \\\\ \\hline\n",
      "\\makecell{IgM \\\\ (g/L)}                       &             0.06$\\pm$0.23 &                 0.05$\\pm$0.21 &                    0.05$\\pm$0.22 &              0.04$\\pm$0.19 &                  0.06$\\pm$0.24 \\\\ \\hline\n",
      "\\makecell{IgA \\\\ (g/L)}                       &             0.17$\\pm$0.37 &                 0.18$\\pm$0.38 &                    0.19$\\pm$0.40 &              0.18$\\pm$0.38 &                  0.18$\\pm$0.39 \\\\ \\hline\n",
      "\\makecell{$\\kappa$-Light Chain \\\\ (mg/dL)}    &             0.64$\\pm$0.48 &                 0.64$\\pm$0.48 &                    0.64$\\pm$0.48 &              0.64$\\pm$0.48 &                  0.65$\\pm$0.48 \\\\ \\hline\n",
      "\\makecell{$\\lambda$-Light Chain \\\\ (mg/dL)}   &             0.34$\\pm$0.47 &                 0.35$\\pm$0.48 &                    0.34$\\pm$0.48 &              0.36$\\pm$0.48 &                  0.33$\\pm$0.47 \\\\ \\hline\n",
      "\\makecell{$\\beta_2$-Microglobulin \\\\ (mg/mL)} &             4.85$\\pm$3.92 &                 4.91$\\pm$3.88 &                    5.13$\\pm$3.98 &              4.45$\\pm$3.68 &                  5.48$\\pm$4.04 \\\\ \\hline\n",
      "\\makecell{M-protein \\\\ (g/dL)}                &             2.81$\\pm$1.81 &                 2.85$\\pm$1.77 &                    2.80$\\pm$1.80 &              2.73$\\pm$1.79 &                  2.99$\\pm$1.75 \\\\ \\hline\n",
      "\\makecell{Creatinine \\\\ ($\\mu$mol/L)}         &          113.94$\\pm$74.25 &              112.90$\\pm$72.86 &                 118.21$\\pm$79.43 &           105.25$\\pm$68.38 &               122.24$\\pm$77.09 \\\\ \\hline\n",
      "\\makecell{Albumin \\\\ (g/L)}                   &            35.96$\\pm$6.23 &                35.93$\\pm$6.27 &                   35.93$\\pm$6.37 &             36.93$\\pm$6.18 &                 34.70$\\pm$6.18 \\\\ \\hline\n",
      "PFS (months)                                  &            16.29$\\pm$9.60 &                16.04$\\pm$9.31 &                   16.91$\\pm$9.02 &             18.95$\\pm$8.89 &                 12.48$\\pm$8.56 \\\\ \\hline\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(cohort_str_df.T.to_latex(escape=False, column_format='cccccc').replace(\"\\\\\\n\", \"\\\\ \\hline\\n\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "color-vegetarian",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "correct-retention",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
